### ISSUE TITLE
error: getting `STATUS_ACCESS_VIOLATION`  in `Device::list()`

### ISSUE THREAD

**Issue by poszu** [cogciprocate/ocl#219]
The following simple program crashes on Windows 11 (Dell XPS 15 with Nvidia 3050 Ti):
```rust
fn main() {
    let platforms = ocl::Platform::list();
    for platform in platforms {
        _ = ocl::Device::list(platform, None).unwrap();
    }
}
```

Error:
```
error: process didn't exit successfully: `target\debug\crash_repro.exe` (exit code: 0xc0000005, STATUS_ACCESS_VIOLATION)
```

I'm unsure how to debug this, I suspect something is wrong with my environment/drivers, etc. Could you please help?

**Comment by c0gent** [comment]
I don't have any Nvidia machines set up right now and am unable to reproduce. 

One thing I like to do is to write a C++ equivalent program to isolate the problem and determine whether it's a driver bug or not. It almost always is. By digging through the three abstraction layers  (ocl -> ocl-core -> cl_sys) you can figure out exactly which parameters with and in what order each OpenCL function is being called. You could also just copy most of the boilerplate  from a tutorial online (Here's one I randomly googled: https://www.eriksmistad.no/getting-started-with-opencl-and-gpu-computing/).

**Comment by c0gent** [comment]
Very interesting. Yeah, if you get time let's try the C++ equivalent just to isolate the problem for sure.

**Comment by poszu** [comment]
Reproduced it with C++. I followed the guide here: https://github.com/KhronosGroup/OpenCL-Guide/blob/main/chapters/getting_started_windows.md.

It seems to crash because `platform->dispatch` is not initialized:
![image](https://github.com/cogciprocate/ocl/assets/3966964/ca43f9ac-6c25-4722-ba02-c8e994984b5d)


The program:
```C++
#include <stdio.h>
#include <vector>

#include <CL/cl.h>

int main()
{
    cl_uint numPlatforms = 0;
    cl_int CL_err = clGetPlatformIDs(0, NULL, &numPlatforms);
    if (CL_err == CL_SUCCESS) {
        printf_s("%u platform(s) found\n", numPlatforms);
    } else {
        printf_s("clGetPlatformIDs(%i)\n", CL_err);
        return 1;
    }

    // Get all platforms
    std::vector<cl_platform_id> platform(numPlatforms);
    CL_err = clGetPlatformIDs( numPlatforms, platform.data(), NULL );
    if (CL_err != CL_SUCCESS) {
        printf_s("clGetPlatformIDs failed: %i\n", CL_err);
        return 1;
    }

    // list names of all platforms
    for (cl_uint id = 0; id < numPlatforms; id++ ) {
        size_t result_size = 0;
        CL_err = clGetPlatformInfo(platform[id], CL_PLATFORM_NAME, 0, NULL, &result_size);
        if (CL_err != CL_SUCCESS) {
            printf_s("clGetPlatformIDs failed: %i\n", CL_err);
            return 1;
        }

        std::vector<char> result(result_size);
        CL_err = clGetPlatformInfo(platform[id], CL_PLATFORM_NAME, result_size, result.data(), NULL);
        if (CL_err != CL_SUCCESS) {
            printf_s("clGetPlatformIDs failed: %i\n", CL_err);
            return 1;
        }

        printf_s("Platform %i: %s\n", id, result.data());
    }
    return 0;
}
```

**Comment by c0gent** [comment]
Presumably some issue with the ICD which is a component which chooses between and loads the actual drivers. I might first try doing a clean install of your Nvidia drivers. If that doesn't work, try installing the latest [Intel OpenCL CPU Runtime](https://github.com/intel/llvm/releases/tag/2022-09) 
(Windows: [win-oclcpuexp-2022.14.8.0.04_rel.zip](https://github.com/intel/llvm/releases/download/2022-09/win-oclcpuexp-2022.14.8.0.04_rel.zip)). Perhaps the Intel ICD will work. 

Beyond that you'll have to do some more digging on troubleshooting the ICD which is something I've never done on Windows and wouldn't be of any help. Good luck!

**Comment by c0gent** [comment]
Here's the Khronos ICD loader page you could try too: https://github.com/KhronosGroup/OpenCL-ICD-Loader. Should work with any driver.

**Comment by c0gent** [comment]
Great!

I'm curious to know how you found your way to this solution. If you have time, would you mind posting how you troubleshooted this for the benefit of anyone else that comes across a similar issue.

No obligation. Grats again :)

**Comment by poszu** [comment]
Sure, but there was nothing more than luck :)

I ran out of options for debugging the C++ program and I figured that the problem must be in the DLL. So I looked over the Internet to see how the DLL is picked up and found a forum post mentioning that registry entry.

It is a chocolatey  [opencl-intel-cpu-runtime](https://community.chocolatey.org/packages/opencl-intel-cpu-runtime) package that installs the DLL and adds this registry entry. Unfortunately, removing the package doesn't remove them - they must be removed manually. My guess is that a DLL meant for the CPU is loaded for the Intel GPU.
